### ISSUE TITLE
User-specified dependencies configuration and upload

### ISSUE THREAD

**Issue by zlav** [fossas/spectrometer#204]

**Comment by scruffystuffs** [comment]
I'm not sure that reading from a config file actually makes sense here.  I'm not suggesting we throw away the config file idea, but I think that this feature should be orthogonal to the config file.  I'd much rather see users select a file on the command line (or pipe it in through standard input) specifically for this info alone.  This would also mean that users wouldn't need a config file for this feature to work.

I am also wary of using YAML for this [(similar concern here)](https://github.com/fossas/spectrometer/issues/203#issuecomment-791700133).  I think JSON is a lot more portable, and could allow people to just bypass us and run straight through the API, if they wanted to (without having to hook up JSON and YAML outputters for their tools).

**Comment by liftM** [comment]
Maybe `additionalDependencies`?

**Comment by zlav** [comment]
Talked with Wes about his concerns here. Major ones listed below:
- If we put the user-specified deps in the config file we are marrying ourselves to that format
- It can be difficult to create new files in CI tools, in the common use case where someone wants to programmatically modify the list of specified dependencies we may be adding complexity.

I think the concern about creating new files in CI is important. We should consider how difficult it would be to allow users to pipe command output into the CLI for specified dependencies, and even if this is not built to start we need to make sure we don't put ourselves in a hole.

The biggest concern that I sympathize with is how to programmatically add dependencies to a configuration file and I think it is worth considering separating the two. 

I do not like the idea of having another hidden file that is picked up alongside configuration. I feel strongly that if we do separate the two, `addditionalDependencies` is explicitly set as a flag or in the configuration file.

**Comment by liftM** [comment]
Hmm, we might also want to have like a `type: csv` that lets you point us to a CSV with `(name, version, license)` tuples for convenience (e.g. if "upload this custom build tool's exported CSV" becomes a common use case).

**Comment by zlav** [comment]
@liftM yea I agree with you. We discussed creating a gist for this use case, but honestly, a flag wouldn't be that much more difficult.

**Comment by cnr** [comment]
Like wes, I don't want to marry this to the project-level config format we settle on.

Instead, I'd like to allow users to create a specifically-named file, e.g., `fossa-deps.json`, that we'd look at during `fossa analyze`. This file could be placed anywhere in the filetree, which would potentially allow for several `fossa-deps.json` to be present:

```
analysis-root/fossa-deps.json
analysis-root/foo/fossa-deps.json
analysis-root/bar/fossa-deps.json
```

---

For the file format: I agree yaml's maybe not the best. I suspect some users would want to use it anyway (especially if they're generating these documents by hand). We could hypothetically support several file formats with pretty minimal overhead (`json`, `yaml`, `toml`)

---

As for the specific shape of dependencies in the file: I generally like the format you have there, with some minor nits:

- I think we should make the use of srclib-style locators explicit with a `type: locator`, which will make everything feel more uniform; and
- The `url` type should be a url locator instead
- Rename "location" to "path" in vendored dependencies, because it's more specific
- Rename "revision" to "version"? I suspect users will incorrectly use "version" a bunch, because it's the more-common terminology, even if we use "revision" internally

```yaml
dependencies:
- type: locator
  fetcher: gem
  package: iron
  version: 1.1.1
- type: locator
  fetcher: mvn
  package: com.google.guava:guava
- type: locator
  fetcher: url
  package: fossa.com/project.tar
- type: vendored
  package: iron
  version: 1.1.1
  path: vendor/fossas
```

**Comment by liftM** [comment]
I like @cnr's suggestions on the shape of dependencies, except `type: locator`. This is more technically correct, but doesn't mean anything to users. We should always prefer exposing concepts in ways that _make intuitive sense_ to users, and then translate them into FOSSA concepts within our backend logic. That is, unless you particularly enjoy explaining what a "locator" is to every customer and every sales engineer because none of them will read our documentation and all of them will be confused and unwilling to write their own configuration file without one of us holding their hands because they're not sure what "locators" are.

----

Re: separating dependencies into a separate file: `fossa-deps.json` just adds another file that users need to create.

> If we put the user-specified deps in the config file we are marrying ourselves to that format

This is a weak argument.

Putting dependencies into the configuration file doesn't "marry us to that format" any more than supporting any kind of dependency file does. We can always evolve the configuration file format - that's the whole point of the "version" field.

> It can be difficult to create new files in CI tools, in the common use case where someone wants to programmatically modify the list of specified dependencies we may be adding complexity.

This is the stronger argument against "deps in config files". I think this is mitigated with a `type: csv` or `type: manifest` specified dependency that points to a programmatically generated file.

Think about this from the perspective of who is actually going to use this feature. It's usually going to be sales engineers manually trying to hack a PoC, or customers specifying an archive upload. The programmatically generated file is the _less common_ use case. We should certainly _enable_ that use case, but we should optimize the experience for the common case.

**Comment by cnr** [comment]
> This is more technically correct, but doesn't mean anything to users
I don't care what we call it. I think it's semantically clearer, at least to me, to use a specific "locator" (or whatever we call it) type to mean "we're going to try to download this from the internet", rather than overloading gem/mvn/git/etc to mean that 

I also suspect we're going to eventually define real buildtool types, e.g., "cocoapods" whenever we're able to actually sum type our dependency representation with more-rich information. As a motivating use-case, cocoapods dependencies are actually a pointer to two things: a registry repo (usually just github/Cocoapods/Specs, but there are other public registry repos) and a package within that repo. We're not able to convey this information with current locators

---

I'm not sure how to feel about additional deps in the configuration file:
- I agree that putting them in the config file probably entails less friction for folks using the feature
- It's less correct in the technical sense, because the concerns are orthogonal (project-wide configuration vs adding dependencies)
- Using a project-wide configuration file isn't the default, so many rollouts of this feature are going to be with customers that aren't already using a configuration file -- there isn't "another file" they need to create; just "a file"
- There have been use-cases with v1 in the past (not the norm, to be fair) where automagically picking up hypothetical `fossa-deps.json` files in the filetree would've made things much easier; instead we were stuck trying to generate one huge `.fossa.yml` via a script at the top level. requiring `type: manifest` or `type: csv` pointers in the project-level config is likely to present similar pains for these use-cases

**Comment by liftM** [comment]
> I don't care what we call it. I think it's semantically clearer, at least to me, to use a specific "locator" (or whatever we call it) type to mean "we're going to try to download this from the internet", rather than overloading gem/mvn/git/etc to mean that
>
> I also suspect we're going to eventually define real buildtool types, e.g., "cocoapods" whenever we're able to actually sum type our dependency representation with more-rich information. As a motivating use-case, cocoapods dependencies are actually a pointer to two things: a registry repo (usually just github/Cocoapods/Specs, but there are other public registry repos) and a package within that repo. We're not able to convey this information with current locators

Yeah, having more expressive options to specify dependencies would be great. But for now, while we're stuck with locators, let's make specifying locators more _user_-friendly (as opposed to _FOSSA_-friendly).

I would be swayed by a compelling argument that either (1) users understand what "locator" means, or (2) it's more important to be clear for internal developers rather than users; but I don't think either statement is true.

----

> Using a project-wide configuration file isn't the default, so many rollouts of this feature are going to be with customers that aren't already using a configuration file -- there isn't "another file" they need to create; just "a file"

This is true - I've seen many cases where someone would have benefited from a dependencies file but wouldn't have needed other fields in the configuration file. But in those cases, they could also just omit the other fields from the configuration file - putting dependencies in the configuration file is no better, but also no worse.

> There have been use-cases with v1 in the past (not the norm, to be fair) where automagically picking up hypothetical fossa-deps.json files in the filetree would've made things much easier; instead we were stuck trying to generate one huge .fossa.yml via a script at the top level. 

This is really interesting. Can you give an example?

> additional deps in the configuration file

The above two arguments are convincing enough that I'm now ambivalent between dependencies in the configuration file or a separate file. I'm open to trying either, and I don't think this is a particularly important or irreversible decision (we can reverse "deps in config" with a new config file version, and we can reverse "deps in separate file" with a script that migrates `fossa-deps.json` files in a file tree into a generated config file).

@cnr if you have a compelling use case for "pick up all `fossa-deps.json` in a file tree, that would be interesting enough for me to lean towards trying the separate file approach.

In either case, this is @zlav's call.

----

I've written about my thoughts about TOML in #203, but if we can support multiple formats cheaply then :shrug: why not, I suppose.

I'm still not sure I'm convinced that a separate deps file is a good idea, but .
